### ISSUE TITLE
Expose parse/parseWithoutProcessing in the IHandlebars interface

### ISSUE THREAD

**Issue by bags1976** [zjklee/Handlebars.CSharp#27]
This would be very useful to allow a user to extract the tokens from a template and then be able to extract the variables required for a particular template.  The feature of extracting the variables was requested long ago here in the native handlebars.js and ultimately is what I would love to see implemented as well. 

https://github.com/handlebars-lang/handlebars.js/issues/1207

Currently there is no way to have the compiler parse the template for tokens and simply return the results in the interface.  The compiler and applicable helpers are internal classes / methods so there is no good way to extract this information now.

Alternatively, simply exposing all the variable names required by a template would be the best but this is not currently available in the handlerbars lang.

The handlebars api has methods to parse and parseWithoutProcessing returning the ast/tokens.  

https://github.com/handlebars-lang/handlebars.js/blob/master/docs/compiler-api.md

**Comment by zjklee** [comment]
If you need only variable names without helper names, you can utilize existing functionality. It's available through `IFeature` interface. `ClosureFeature` is build-in feature and is responsible for extracting constants from expression tree before it's been compiled. `ClosureFeature.TemplateClosure.Store` would contain all extracted constants.
`PathInfo` struct is responsible for storing parsed path expressions from a template. Unfortunately this is an internal type. However you can access the path value by `.ToString()`.

Consider the following example:
```csharp
[Fact]
public void AccessVariables()
{
    var source = "{{helper1 variable1.path1}} + {{helper2 variable2.[$path1].path2}}";
    var handlebars = Handlebars.Create();

    var variables = new List<string>();
    var accessFeatureFactory = new VariablesAccessFeature(variables);
    handlebars.Configuration.CompileTimeConfiguration.Features.Add(accessFeatureFactory);
    _ = handlebars.Compile(source);
    handlebars.Configuration.CompileTimeConfiguration.Features.Remove(accessFeatureFactory);
    
    Assert.Collection(variables,
        s => Assert.Equal("variable1.path1", s),
        s => Assert.Equal("variable2.[$path1].path2", s)
    );
}

public class VariablesAccessFeature : IFeature, IFeatureFactory
{
    private readonly ICollection<string> _variables;
    private ICompiledHandlebarsConfiguration _configuration;

    public VariablesAccessFeature(ICollection<string> variables) => _variables = variables;

    public IFeature CreateFeature() => this;
    
    public void OnCompiling(ICompiledHandlebarsConfiguration configuration) => _configuration = configuration;

    public void CompilationCompleted()
    {
        var closureFeature = _configuration.Features.OfType<ClosureFeature>().Single();
        var structs = closureFeature.TemplateClosure.Store.Where(o =>
        {
            var type = o.GetType();
            return type.IsValueType && !type.IsPrimitive && !type.IsPublic;
        });

        foreach (var @struct in structs)  _variables.Add(@struct.ToString());
    }
}
```

The proposed approach does not cover whole functionality from handlebarsjs but can be used as a workaround if you need just variable names.

In case this is not enough I'd definitely be happy to accept a PR covering missing parts.

Good place to start investigation could be the following line: [source/Handlebars/Compiler/HandlebarsCompiler.cs#L31](https://github.com/zjklee/Handlebars.CSharp/blob/master/source/Handlebars/Compiler/HandlebarsCompiler.cs#L31) - after the line parsing is done and next step is processing and compilation.

**Comment by zjklee** [comment]
In addition to my previous reply.
PR zjklee/Handlebars.CSharp/pull/21 would make some of the internal types public (e.g. `PathInfo`) so it should simplify an example above.

**Comment by bags1976** [comment]
That workaround you provided was exactly what I needed for my use case thank you.  

I did take a look at implementing the full parse method as well and it could be partially implemented fairly easily by converting the tokens into public Statements per the handlebars interface specs but it does not have the start and end locations only the value and type.

**Comment by zjklee** [comment]
After /pull/21 is merged my previous example can to updated.
Notes:
- `PathExpressionCollectorFeature` - collects all path expressions ("variables" and helpers)
- `VariableCollectorFeature` - collect only "variables"

```csharp
[Fact]
public void AccessVariables()
{
    var source = "{{helper1 variable1.path1}} + {{helper2 variable2.[$path1].path2}}";
    var handlebars = Handlebars.Create();

    var variables = new List<PathInfo>();
    var pathExpressions = new List<PathInfo>();
    var variablesCollectorFeature = new VariableCollectorFeature(variables);
    var pathExpressionCollectorFeature = new PathExpressionCollectorFeature(pathExpressions);
    handlebars.Configuration.CompileTimeConfiguration.Features.Add(variablesCollectorFeature);
    handlebars.Configuration.CompileTimeConfiguration.Features.Add(pathExpressionCollectorFeature);
    
    _ = handlebars.Compile(source);
    
    handlebars.Configuration.CompileTimeConfiguration.Features.Remove(pathExpressionCollectorFeature);
    handlebars.Configuration.CompileTimeConfiguration.Features.Remove(variablesCollectorFeature);

    Assert.Contains(pathExpressions, info => info == "helper1");
    Assert.Contains(pathExpressions, info => info == "helper2");
    Assert.Contains(pathExpressions, info => info == "variable1.path1");
    Assert.Contains(pathExpressions, info => info == "variable2.[$path1].path2");
    
    Assert.DoesNotContain(variables, info => info == "helper1");
    Assert.DoesNotContain(variables, info => info == "helper2");
    Assert.Contains(variables, info => info == "variable1.path1");
    Assert.Contains(variables, info => info == "variable2.[$path1].path2");
}

public class PathExpressionCollectorFeature : IFeature, IFeatureFactory
{
    private readonly ICollection<PathInfo> _variables;
    private ICompiledHandlebarsConfiguration _configuration;

    public PathExpressionCollectorFeature(ICollection<PathInfo> variables) => _variables = variables;

    public IFeature CreateFeature() => this;

    public void OnCompiling(ICompiledHandlebarsConfiguration configuration) => _configuration = configuration;

    public void CompilationCompleted()
    {
        foreach (var pathInfo in _configuration.PathInfoStore.Values) _variables.Add(pathInfo);
    }
}

public class VariableCollectorFeature : IFeature, IFeatureFactory
{
    private readonly ICollection<PathInfo> _variables;
    private ICompiledHandlebarsConfiguration _configuration;

    public VariableCollectorFeature(ICollection<PathInfo> variables) => _variables = variables;

    public IFeature CreateFeature() => this;

    public void OnCompiling(ICompiledHandlebarsConfiguration configuration) => _configuration = configuration;

    public void CompilationCompleted()
    {
        var pathInfos = _configuration.Features
            .OfType<ClosureFeature>()
            .Single().TemplateClosure.Store
            .OfType<PathInfo>();

        foreach (var pathInfo in pathInfos) _variables.Add(pathInfo);
    }
}
```

**Comment by zjklee** [comment]
> @bags1976 
I did take a look at implementing the full parse method as well and it could be partially implemented fairly easily by converting the tokens into public Statements per the handlebars interface specs but it does not have the start and end locations only the value and type.

Unfortunately parse is not really good at determining `start` and `end` positions. I did some work to improve the situation (see [`IReaderContext`](https://github.com/zjklee/Handlebars.CSharp/blob/master/source/Handlebars/IO/IReaderContext.cs)) but it only supports `start` position and currently used to report better errors.
I considered moving to parser combinators (zjklee/Handlebars.CSharp/issues/16) but it looks to be an unjustified effort.
